Alpha light

Tixxx opened the pull request.

vaeksare committed:
Squashed commit of the following:

commit 44fd7f8b89b350bcb68a2b9e58f9ca045cee7bbf
Merge: a3d5910 17e8d9c
Author: Tixxx <tix@microsoft.com>
Date:   Thu Sep 5 14:34:51 2019 -0700

    Merge pull request #11 from Tixxx/saemal/msallreducecudakernels

    Saemal/msallreducecudakernels

commit 17e8d9c5b63a153465a900b9beab4cbbb444c56c
Merge: 03e225d a3d5910
Author: Saeed Maleki <30272783+saeedmaleki@users.noreply.github.com>
Date:   Wed Sep 4 15:55:17 2019 -0700

    Merge branch 'tix/vhddwithlocalreduction' into saemal/msallreducecudakernels

commit 03e225d6fb403b42d140a01b84a94de19d7537b7
Author: Ubuntu <ubuntu@ip-172-31-4-98.us-west-2.compute.internal>
Date:   Wed Sep 4 22:35:16 2019 +0000

    tested ring allreduce for msallreduce

commit 66305faebd7702b15f9aac4b10b16a57e0e18cfa
Author: Ubuntu <ubuntu@ip-172-31-4-98.us-west-2.compute.internal>
Date:   Wed Sep 4 01:36:39 2019 +0000

    fixed the ring order

commit 9331635ee2a0bbd43effb2b40d37b24746ef8c42
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Fri Aug 30 20:40:28 2019 +0000

    fixed most bugs

commit a15ec1d5f8f3030b461adf35b9a1d1f5de29dcf6
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Tue Aug 27 19:58:29 2019 +0000

    checking before the nd40 goes away

commit a3d59102865f3127a48ed41fd80f5a94d0bc988d
Author: Tix <tix@microsoft.com>
Date:   Tue Aug 27 11:19:12 2019 -0700

    changed init and finalize logic in ms_cuda_msallreduce

commit cd4aaedf0ad5e0097db915f79012f07c5aa01928
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Mon Aug 26 22:53:07 2019 +0000

    testing the ring allreduce

commit 254cd7ff16f401f51ac8670e86fe38219c29090d
Merge: d485099 e74f098
Author: Tixxx <tix@microsoft.com>
Date:   Mon Aug 26 12:30:22 2019 -0700

    Merge pull request #10 from Tixxx/saemal/kernelcallsformsallreduce

    Saemal/kernelcallsformsallreduce

commit e74f0988c1e4e5de4fadc0898e48942f51ebbc9f
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 26 12:04:29 2019 -0700

    fixed copying from device to host

commit fc4c733cd80647f429bf3739f80205b14a73f370
Merge: d485099 4491b32
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 26 11:00:27 2019 -0700

    Merge branch 'saemal/kernelcallsformsallreduce' of https://github.com/Tixxx/horovod into saemal/kernelcallsformsallreduce

commit f518e95b871586b4af17a6eac6bb8c510c3e6066
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Fri Aug 23 22:52:34 2019 +0000

    merged with ring allreducew

commit e8bcec9a8d1675b858aef41ea33ee540d30bfffe
Merge: 4491b32 45b3488
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Fri Aug 23 21:38:06 2019 +0000

    Merge branch 'olsaarik/ringplusvhdd' into saemal/msallreducecudakernels

commit 4491b327902a24a6aed99ff9c2c9917650d824d5
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Fri Aug 23 21:32:20 2019 +0000

    fixed bug in setup.py

commit 45b3488114ed714945b2aa508bc739eb90ae6dc2
Author: Olli Saarikivi <olsaarik@microsoft.com>
Date:   Fri Aug 23 21:28:38 2019 +0000

    Fix variable declarations

commit a1093e24ddd545efa111554bbca7f55bc6a7b4ed
Author: Olli Saarikivi <olsaarik@microsoft.com>
Date:   Fri Aug 23 21:11:50 2019 +0000

    Set ring cuda msallreduce as default

commit eda4e4ecf07bf9f518136fd5364623e75731033e
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Fri Aug 23 18:20:20 2019 +0000

    cuda kernels compiles now -- need to fix for -arch=sm_ <60

commit 84288ad8953624cd1e436bc646312ed5134dce6d
Author: Olli Saarikivi <olsaarik@microsoft.com>
Date:   Fri Aug 23 17:54:01 2019 +0000

    Add hierarchical ring vhdd msallreduce

commit d485099b1e2de961cbd8e12c2e74c842500604ea
Author: Tix <tix@microsoft.com>
Date:   Fri Aug 23 06:33:40 2019 -0700

    fixed a type error in msallreduce

commit 66049004e9b9ac3b524d94620c5cad79c77f8558
Merge: 71a82d9 2595113
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Thu Aug 22 18:44:20 2019 +0000

    Merge branch 'saemal/msallreducecudakernels' of https://github.com/Tixxx/horovod into saemal/msallreducecudakernels

commit 71a82d99e20a3507c1b2ef5b65f9ab3dccca87fb
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Thu Aug 22 18:44:19 2019 +0000

    fixing bugs with setup.py

commit 2595113312e3470cdef04266712867e464bd9a0d
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Thu Aug 22 18:42:44 2019 +0000

    added the CMakeList file for cuda kernel

commit 799fc47e39090f2e5e6c797a65b780a3b19c6b64
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Thu Aug 22 07:36:32 2019 +0000

    cuda kernel compiles now

commit 925d3e4325a87aebd13141dc9aaad1627c87ae38
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Tue Aug 20 17:29:53 2019 -0700

    added kernel calls and the hooks for calling them

commit e69452a3135c64455f129550312a047bf468da4e
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Tue Aug 20 17:29:21 2019 -0700

    added kernel calls and the hooks for calling them

commit d6408c9cfb29ff6f220be283ac2403b524c65077
Author: Tix <tix@microsoft.com>
Date:   Tue Aug 20 14:56:46 2019 -0700

    fixed correctness bug

commit eabaa57a178aad81c91ab20df6e387d27e09a9f0
Merge: 4245b57 75363ef
Author: Tixxx <tix@microsoft.com>
Date:   Fri Aug 16 09:39:46 2019 -0700

    Merge pull request #7 from Tixxx/tix/vhddwithlocalreductiongpu

    tixTix/vhddwithlocalreductiongpu

commit 75363efe120193f840c166eb6ca3143790a928cf
Author: Tix <tix@microsoft.com>
Date:   Fri Aug 16 09:26:29 2019 -0700

    PR comments
    assign streams based on layerid and number of threads.
    Name change for cublas initilization method

commit e3c75f7d482a8d0045971667d968a05e71610526
Author: Tix <tix@microsoft.com>
Date:   Thu Aug 15 17:18:43 2019 -0700

    fixed mem leak.
    fixed seg fault.
    improved stream usage.

commit da32b1fa17b37f8b11b885eae8b41077442c8294
Author: Tix <tix@microsoft.com>
Date:   Thu Aug 15 01:27:02 2019 -0700

    fixed multithreading issue with tensorflow
    give each thread a cuda stream
    fixed communicator bug caused by merge

commit 30056aa0357465e8abaca79ba32b421a134b208a
Merge: 756b4fa 4245b57
Author: Tix <tix@microsoft.com>
Date:   Wed Aug 14 23:48:56 2019 -0700

    Merge branch 'tix/vhddwithlocalreduction' of https://github.com/Tixxx/horovod into tix/vhddwithlocalreductiongpu

commit 756b4fae32a2d2e98698d29d3ebd1c2965332eb2
Author: Tix <tix@microsoft.com>
Date:   Wed Aug 14 22:48:00 2019 -0700

    added fp16 support for gpu

commit 4245b57a0087fa1d785edde739db831ba1928067
Merge: 2a1eedf 04fa0e4
Author: klipto <todd.mytkowicz@gmail.com>
Date:   Wed Aug 14 17:17:11 2019 -0700

    Merge pull request #9 from Tixxx/tree_local_reduce

    tree local reduce

commit 04fa0e404e14b60d721bf0c3420922aab2be5fea
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Thu Aug 15 00:15:39 2019 +0000

    simple fix

commit 1f5c22f7ae982fd4bbf08e4ced2ecedb116ac409
Author: Saeed Maleki <saemal@microsoft.com>
Date:   Wed Aug 14 23:58:15 2019 +0000

    tree local reduce

commit 33dbe83d6d09aa127f76c2690a4f192ed7bfc19f
Author: Tix <tix@microsoft.com>
Date:   Tue Aug 13 15:56:53 2019 -0700

    fixed cuda init to make gpu reduction work

commit 93d7b3796b09dc276271acf51e1872c13a2f9ce4
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 12 15:37:14 2019 -0700

    addressed some comments in pr

commit bc889f367cb167ed201ece6ba9dd643b2157b874
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 12 14:19:46 2019 -0700

    integration branch

commit 68de8a1097f31e4b0551fa4615243fd032d0290d
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 12 14:18:09 2019 -0700

    changed to cublasxxxEx call and only with float32

commit 831297617b022284ccd99b3cb4f4f28d8ec3ea6e
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 12 13:29:42 2019 -0700

    compile pass.
    divide by zero exception in float to double casting

commit 505aed1866e77a62c7b6961037ffb8d6f8a77d8f
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 12 10:42:26 2019 -0700

    adding gpu support for ms allreduce logic
    in progress

commit 2a1eedf828b3eddbe4d2b5ae22579fc67ef21755
Merge: a1913e8 d33fa92
Author: Vadim Eksarevskiy <42353187+vaeksare@users.noreply.github.com>
Date:   Fri Aug 9 15:57:29 2019 -0700

    Merge pull request #5 from vaeksare/vaeksare/separate_average

    Vaeksare/separate average

commit d33fa9278c29f90112cc868fc9923436d34c437f
Author: Vadim Eksarevskiy <vaeksare@microsoft.com>
Date:   Fri Aug 9 14:54:15 2019 -0700

    deleted accidental binary files

commit 2e63692047c2b8317533ea0666cbba2d3f56bfb5
Author: Vadim Eksarevskiy <vaeksare@microsoft.com>
Date:   Fri Aug 9 14:51:00 2019 -0700

    refactored msallreduce to be a separate op in horovod

commit a1913e8caa41e259b5486dfad49cc3ba5c3751b8
Merge: 3a8cdd2 9accd83
Author: klipto <toddm@microsoft.com>
Date:   Fri Aug 9 14:15:47 2019 -0700

    Merge branch 'tix/vhddwithlocalreduction' of https://github.com/Tixxx/horovod into tix/vhddwithlocalreduction

commit 3a8cdd26ae6038026dd6f2f7dac6bb77332c1c4f
Author: klipto <toddm@microsoft.com>
Date:   Fri Aug 9 14:06:02 2019 -0700

    workaround for # of elements/size issue

commit 55e6ce18cd6e00cc81c7ad785a713d0506dcaea8
Author: root <root@GCRHYPCBJ016.redmond.corp.microsoft.com>
Date:   Fri Aug 9 13:29:42 2019 -0700

    fixed load and added guard for potential bug

commit 9accd83e62507fed7b1d58d73062a4980e6fab2c
Author: Tix <tix@microsoft.com>
Date:   Fri Aug 9 11:28:48 2019 -0700

    simplified average logic

commit e364f14f6ef86d633835cd25ffb9ebb2b454ab15
Merge: 278e86c 3dde0e4
Author: Tix <tix@microsoft.com>
Date:   Thu Aug 8 10:09:14 2019 -0700

    Merge branch 'tix/vhddwithallreduce' into tix/vhddwithlocalreduction

commit 278e86c54a2c9a6c6200a003f4398f9a7e8bb3be
Author: Tix <tix@microsoft.com>
Date:   Wed Aug 7 17:02:52 2019 -0700

    merge with tf fixes

commit 3dde0e448a53a3c02756f2457c2d6c5b460d2663
Merge: 83e68e1 a0b9469
Author: klipto <todd.mytkowicz@gmail.com>
Date:   Wed Aug 7 16:32:43 2019 -0700

    Merge pull request #4 from Tixxx/adding_test_functionality

    Added a test for fp16,32,64 tensor allreduce correctness

commit a0b9469fac2aa885da78854de4db284ee16a2464
Author: Todd Mytkowicz <toddm@microsoft.com>
Date:   Wed Aug 7 13:52:44 2019 -0700

    Added a test for fp16,32,64 tensor allreduce correctness

commit 83e68e1a925d98e5cf2a4f78f17186eec8a3e489
Author: Tix <tix@microsoft.com>
Date:   Wed Aug 7 13:33:47 2019 -0700

    replaced local reduction with mpi allreduce

commit c1e5f9cdad903869b9ecfc955bc27bd8a818f1fb
Author: Tix <tix@microsoft.com>
Date:   Tue Aug 6 14:34:56 2019 -0700

    added more optimization flags for compiler

commit 5509baf82dec862d48a4575f1f57310f2c383de7
Author: Tix <tix@microsoft.com>
Date:   Tue Aug 6 09:29:21 2019 -0700

    integrated with the vhdd bug fix

commit dfda595b7314eb184597e792f08b372109eb6db4
Merge: c3c0257 efe1886
Author: Vadim Eksarevskiy <42353187+vaeksare@users.noreply.github.com>
Date:   Mon Aug 5 18:20:30 2019 -0700

    Merge pull request #2 from vaeksare/vaeksare/hvdd

    pytorch workaround

commit efe18862abbf9e3cebcd0442e9f1586e5c93728f
Author: Vadim Eksarevskiy <vaeksare@microsoft.com>
Date:   Mon Aug 5 18:18:19 2019 -0700

    pytorch workaround

commit c3c0257c888e262ccb5a7165406f5964360aab23
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 5 17:50:39 2019 -0700

    merged with vhdd.
    merged with fix in TF averaging logic.

commit b02994aeff09bb25947cfade4a7a4bba0f33192d
Author: Tix <tix@microsoft.com>
Date:   Mon Aug 5 11:37:23 2019 -0700

    added float16 data type

commit 6116e7e4a7183fa281267e01e6a6f4dcca5afb8c
Author: Tix <tix@microsoft.com>
Date:   Fri Aug 2 18:44:20 2019 -0700

    fixed averaging bug in tensorflow

commit b8cab292cd814c308b81a2865bd301b7b2972405
Author: Tix <tix@microsoft.com>
Date:   Thu Aug 1 14:29:56 2019 -0700

    added new parasail algo

commit fa658eb92957cd3d8e0c549265193ae1316353bf
Author: Tix <tix@microsoft.com>
Date:   Thu Aug 1 09:37:34 2019 -0700

    integrated new parasail algorithm

commit 4402dac3f2ee1e8261901cb2f97f91c5876afaa1
Author: Tix <tix@microsoft.com>
Date:   Tue Jul 30 10:43:29 2019 -0700

    added single and multiple large tensor test

commit f6e6c8966bfb27c5ad96e82d47b9fb61cecadafa
Author: Tix <tix@microsoft.com>
Date:   Fri Jul 26 17:22:47 2019 -0700

    merged with local change

commit 6d5fd6c747b82b6f7ef49233a23f85ba72efd112
Author: Tix <tix@microsoft.com>
Date:   Fri Jul 26 17:21:04 2019 -0700

    merged with temp_buffer

commit 46e6ab4c63052f9b68922b9a7885363e6f11c712
Merge: 9c0a7ac cb29e32
Author: Vadim Eksarevskiy <vaeksare@microsoft.com>
Date:   Fri Jul 26 14:34:02 2019 -0700

    fix merge conflict in global state

commit 9c0a7ac7801690b25e8a0a48c15ff05e02304847
Author: Vadim Eksarevskiy <vaeksare@microsoft.com>
Date:   Fri Jul 26 13:44:36 2019 -0700

    added basic pytorch tests for msallreduce

commit c5b1a7fd9fda26d8a0dbb4a495a0681a577e671a
Author: Vadim Eksarevskiy <vaeksare@microsoft.com>
Date:   Thu Jul 25 17:27:22 2019 -0700

    added temp buffer for msallreduce op

commit a7c14a55dc2aa9be60ffc84f7687ca11a25c74d2
Author: Tix <tix@microsoft.com>
Date:   Fri Jul 26 13:52:16 2019 -0700

    fixed some issues with broadcast when fusing respones. Added more logging.

commit cb29e322bde381ed024a235e23c1c90c12032332
Author: Vadim Eksarevskiy <vaeksare@microsoft.com>
Date:   Fri Jul 26 13:44:36 2019 -0700

    added basic pytorch tests for msallreduce

commit bc40e8751e01fec314677d687f5531f508327183
Author: Vadim Eksarevskiy <vaeksare@microsoft.com>
Date:   Thu Jul 25 17:27:22 2019 -0700

    added temp buffer for msallreduce op

commit b644b1b31b2b87ca2fd2d3707ee4022a913e5c4d
Author: Tix <tix@microsoft.com>
Date:   Thu Jul 25 14:01:43 2019 -0700

    fixed seg fault. added multi-tensor test

commit 7babc10db56e17fbb34bcec9e80ce5117bae7ca5
Author: Tix <tix@microsoft.com>
Date:   Wed Jul 24 22:45:52 2019 -0700

    fixed seg fault for 1 tensor case, still happens for multipl tensors

commit 81f4de365ff022280f4fd976fbb34620e4f2b9e1
Author: Tix <tix@microsoft.com>
Date:   Wed Jul 24 13:40:29 2019 -0700

    committing rest of the parallel code. debugging seg fault..

commit 5fadb9ddeb21507c774fd72dce6078463da165f8
Author: Tix <tix@microsoft.com>
Date:   Tue Jul 23 21:50:23 2019 -0700

    incorporated threadpool and changed global state class.
    Added test.

commit 4bf49e6898cc8070baf1930ac36973bd38365004
Author: Tix <tix@microsoft.com>
Date:   Tue Jul 23 14:22:51 2019 -0700

    added more logging and data types for ms allreduce

commit e4e3bb6da49180ac91fd9f315129b7bb2bf5b131
Author: Tix <tix@microsoft.com>
Date:   Tue Jul 16 15:15:47 2019 -0700

    moved p2p comm implementations to header file

commit 730e9fbf22c11fca74a00c95461544ec53c95add
Author: Tix <tix@microsoft.com>
Date:   Tue Jul 16 13:00:36 2019 -0700

    first commit of p2p comm together with parasail op

vaeksare committed:
fix merge conflicts and port over ops.cc

Commit:
fixed gpu build

Commit:
Merge pull request #12 from Tixxx/tix/fixBuild

fixed gpu build

Commit:
first commit of refactoring

Commit:
name change

Commit:
second commit of cleanup:
1. removed template functions, replaced with one template class to contain adasum logic
2. separated mpi from adasum logic. adasum is now independent from comm
3. modularized reduction functions

Commit:
1. fixed allreducetypes in python layer
2. added a function switch in cuda operations

olsaarik committed:
Pull in code for NCCL+Parasail hierarchical op

olsaarik committed:
Dispatch Parasail allreduce call in a separate funcion

olsaarik committed:
Small fixes

olsaarik committed:
Move adasum algo switch to top level in Execute

Various bug fixes

olsaarik committed:
Move adasum to require HOROVOD_GPU_ALLREDUCE=NCCL

This is required for the new NCCL+Adasum hierarchical allreduce.
Alternatively that could be split to a new op.

olsaarik committed:
Fix logic for adasum in TF DistributedOptimizer

Commit:
1. moved mpi related code to correct places
2. factored out ring allreduce functions
3. fixed some bugs related to response cache and mpi calls

olsaarik committed:
Fix bug in adasum NCCL hierarchical

Add tracing prints for which adasum algo is called.

Commit:
Merge branch 'tix/master_rebase_cleanup' of https://github.com/Tixxx/horovod into tix/master_rebase_cleanup

Commit:
guard nccl related code with HAVE_NCCL directive

Commit:
1. added pipeline to ring
2. changed enabled function to return true based on adasum env

olsaarik committed:
Fix logic for adasum in _DistributedGradientTape

Commit:
Addressed PR comments.

Commit:
removed some redundant includes

Commit:
Create rings based on NVlink topology.

Commit:
added a function to check pointer type so that gpu operations can use cpu computations as well.
added a cudaSetDevice call when multi-threading
added ADASUM in response cache

Commit:
Merge pull request #13 from Tixxx/tix/master_rebase_cleanup

Tix/master rebase cleanup

Commit:
Merge branch 'master' into parasail_alpha

Commit:
Removed custom build scripts

Commit:
Merge pull request #15 from kit1980/adasum_alpha

Create rings based on NVlink topology.

olsaarik committed:
Replace adasum nccl+vhdd with smarter algorithm

Algorithm and core implementation courtesy of Saeed Maleki.

olsaarik committed:
Fixes for new nccl+adasum hierarchical

olsaarik committed:
Code cleanup

olsaarik committed:
Merge pull request #16 from olsaarik/smartncclvhddhierarchical

Replace adasum nccl+vhdd with smarter algorithm

Commit:
added average after nccl local reduction

Commit:
debugging

Commit:
divided by element size

Commit:
debugging

Commit:
testing adasum for extreme case of orthogonal tensors

Commit:
Revert "Merge pull request #16 from olsaarik/smartncclvhddhierarchical"

This reverts commit a05e04a5a85b8b30ce40d0483dc2f39bce2aea14, reversing
changes made to 94e0b7c59848af25cb09b95bf32077f72c775de4.

Commit:
revert back to original code

Commit:
1. fixed average when using nccl ring
2. cleaned up env variables

Commit:
testing..

olsaarik committed:
Divide by local_size for adasum GPU_NCCL_LOCAL_AVG

Commit:
added divisor logic to mpi_ops_v2

Commit:
added pytorch test

Commit:
fixed fusion buffer bug when using nccl_local_avg

Commit:
changed test to be more generic

Commit:
added tests
fixed new adasum nccl algorithm bug

Commit:
PR comments
changed pytorch tests

olsaarik committed:
Add op parameter to DistributedOptimizers

Also change allreduce_type to op in allreduce functions. With these
changes using adasum is as easy as:
>hvd.allreduce(tensor, op=hvd.Adasum)
or
>opt = DistributedOptimizer(opt, op=hvd.Adasum)

No longer check for HOROVOD_ADASUM env var in framework code. The intent
is that HOROVOD_ADASUM will only be used to select algorithm
implementations, but we will default to something reasonable without it.
Setting that default is still todo.

Update docstrings to reflect new parameters.

Commit:
PR comments

Commit:
Merge pull request #18 from Tixxx/tix/fixnccl

Tix/fixnccl

olsaarik committed:
Change Adasum env var setup and algo selection

HOROVOD_ADASUM is now HOROVOD_ADASUM_GPU since both CPU and GPU
reductions might have to be made in the same job and there's no choices
for CPU algorithms.

Also if HOROVOD_ADASUM_GPU is not set any GPU reductions default to the
GPU hierarchical (that assumes a Cuda aware MPI).

olsaarik committed:
Merge remote-tracking branch 'tj/adasum_alpha' into olsaarik/optimizerparam

olsaarik committed:
Bugfix

olsaarik committed:
Remove reduce op values from python side

Instead values are read at import time through HorovodBasics class.

olsaarik committed:
Bugfix

Commit:
use forward declaration instead of including common.h

olsaarik committed:
Fix bug with the averaging for NCCL_LOCAL_AVG

If the user had set HOROVOD_ADASUM_GPU=NCCL_LOCAL_AVG, but actually
placed tensors on CPU, then the local averaging would still be applied
in framework code. Now the placement of the tensor to be reduced is
checked.

olsaarik committed:
Bugfix in adasum torch tests

olsaarik committed:
Merge branch 'adasum_alpha' of github.com:Tixxx/horovod into adasum_alpha

olsaarik committed:
Merge branch 'adasum_alpha' into olsaarik/optimizerparam

olsaarik committed:
Bugfix

olsaarik committed:
Add adasum CPU_VHDD algorithm

This does VHDD on all ranks and levels

olsaarik committed:
Fixes to get CPU_VHDD enabled and fusion working

olsaarik committed:
Bugfix

Commit:
p2pallreduce had a bug

olsaarik committed:
Merge branch 'adasum_alpha' into olsaarik/purevhdd

olsaarik committed:
Small fixes

olsaarik committed:
Replace p2pallreduce with normal MPI for fused adasum

olsaarik committed:
Merge pull request #21 from olsaarik/olsaarik/purevhdd

Fused VHDD on CPU on all ranks and levels

olsaarik committed:
Merge branch 'adasum_alpha' into olsaarik/optimizerparam

Commit:
lightweight version of alpha

Commit:
changed test

Commit:
fixed bug with tf.test.is_gpu_available

Commit:
format cleanup

Commit:
deleted submodules

vaeksare commented on horovod/common/ops/mpi_p2p_operations.cc:
```diff
@@ -0,0 +1,148 @@
+
+namespace horovod {
+namespace common {
+AdasumMPIP2pOp::AdasumMPIP2pOp(MPIContext* mpi_context)
```
Is there a good reason for this to be called Adasum and inherit from the AdasumOp? From what I saw, the only function from AdasumOp this uses is GetPerElementSize(), which also does not seem Adasum specific. Seems like this class could be useful for other things in the future. I am guessing the reason behind the inheritance being like this currently:
AdasumMPIP2pOp -> AdasumOp -> P2pOp 
is due to possibly adding support for Gloo in the future? But this still seems strange and possibly redundant. P2pOp and AdasumOp kind of feel like they should be 2 different Ops not directly tied in inheritance. Thoughts on maybe having AdasumOp store a class of P2pOp instead of inheriting from it and use that class's public methods to perform the peer to peer operations?

Tixxx replied:
talked offline. Since the adasumMPIP2p has specific logic to create communicators for vhdd which will be used by both Adasum_cuda and adasum_mpi, the current structure is to avoid code duplication of that. As there wont be other scenarios that will use these p2p operations for now, i will leave the structure as is, will re-visit to see if composition is needed here.

Commit:
pr comments

vaeksare commented on horovod/common/fusion_buffer_manager.h:
```diff
@@ -37,6 +37,8 @@ class FusionBufferManager {
   //  on_start_init: Callback on starting buffer initialization.
   //  on_end_init: Callback on completing buffer initialization.
+  //  validity_check: function to use to compare existing and requested buffer sizes
+  //    to see if the existing buffer can be reused.
```
Forgot to remove this comment.

vaeksare commented on horovod/common/parameter_manager.cc:
```diff
@@ -233,7 +233,6 @@ ParameterManager::Params ParameterManager::GetParams() {
   }
 
   params.active = active_;
-
```
Spacing

vaeksare commented on horovod/common/parameter_manager.h:
```diff
@@ -29,7 +29,6 @@
 
 namespace horovod {
 namespace common {
-
```
Spacing

vaeksare commented on horovod/common/parameter_manager.h:
```diff
@@ -229,6 +228,7 @@ class ParameterManager {
 
   std::vector<ITunableParameter*> parameter_chain_;
   bool active_;
+  
```
Spacing

Commit:
more format changes

jaliyae commented on horovod/common/ops/adasum_cuda_operations.cc:
```diff
@@ -0,0 +1,269 @@
+    buffer_len = num_elements * element_size;
+  }
+
+  // Split the elements into two groups: num_elements_per_rank*local_size,
```
nice comment

olsaarik replied:
not ours, sadly. most of this function is copy-paste from horovod/common/ops/nccl_operations.cc

jaliyae commented on horovod/common/ops/adasum_operations.h:
```diff
@@ -0,0 +1,459 @@
+    int rank = GetLocalRankWithComm(communicator);
+    int size = GetSizeWithComm(communicator);
+
+    if (IsPowerOfTwo(size) == false) {
```
Could we handle this case or put an error in this case?

Tixxx replied:
oh i forgot to change the error message here. i will change it.

jaliyae commented on horovod/tensorflow/__init__.py:
```diff
@@ -47,21 +50,31 @@ def allreduce(tensor, average=True, device_dense='', device_sparse='',
+    true_op = Sum if op == Average else op
+
     if isinstance(tensor, tf.IndexedSlices):
+        # TODO: Need to fix this to actuall call Adasum
```
is this tod valid?

Tixxx replied:
yes, we dont have a good story for sparse tensors now. Original parasail had a python implementation of the tree, with vhdd we have to find a better design to handle this

Commit:
removed FusedVHDD

vaeksare commented on horovod/common/ops/adasum_mpi_operations.cc:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  // Do allreduce.
+  timeline.ActivityStartAll(entries, MPI_ALLREDUCE);
```
Thoughts on creating new Timeline names for these activities and recording this as something like MPI_ADASUM?

olsaarik replied:
I think that would make sense. The timeline's going to be used for performance insights and we have different performance characteristics than MPI_Allreduce

Tixxx replied:
sounds good. i will change it

olsaarik committed:
Guard local averaging for Adasum with nccl_built()

Also move checks for if GPUs are available to the end of the checks,
just in case those checks happen to be slow.

olsaarik commented on horovod/common/ops/adasum_cuda_operations.cc:
```diff
@@ -0,0 +1,269 @@
+					i++;
+				}
+			}
+		}
```
Indentation in broken here

olsaarik committed:
Fix some indentation

olsaarik committed:
Merge pull request #23 from olsaarik/olsaarik/localavgcheckimprovement

Guard local averaging for Adasum with nccl_built()

olsaarik committed:
Rename world_reduction_comms_ to drop world_

Also remove two unused "rank_log_size_" variables.

olsaarik committed:
Merge branch 'alpha_light' of github.com:Tixxx/horovod into alpha_light

olsaarik committed:
Don't drop name in a torch allreduce call

Also don't repeat default parameters in another torch allreduce call.

olsaarik committed:
Small fixes to docstring formatting

olsaarik commented on horovod/common/ops/p2p_operations.h:
```diff
@@ -0,0 +1,49 @@
+
+// Abstract class to expose basic point to point communication interface
+template<typename Communicator_type>
+class PointToPointOp {
```
This class could be merged into AdasumOp, since that's the only class that inherits from this

Tixxx replied:
yea sounds good.

olsaarik commented on horovod/common/ops/adasum_operations.h:
```diff
@@ -0,0 +1,459 @@
+
+// Interface for Adasum algorithm
+template <typename Communicator_type>
+class AdasumOp : public PointToPointOp<Communicator_type> {
```
With the new inheritance hierarchy this class isn't an "Op" anymore in the Horovod sense (doesn't inherit from HorovodOp). Maybe we could rename this to just "adasum.h" (since this file contains the essence of it).

Similarly AdasumMPIP2pOp in mpi_p2p_operations.h/cc isn't an "Op" anymore, so it could also be renamed. For example AdasumMPI in adasum_mpi.h could work. (Also see my other comment concerning p2p_operations.h)

Tixxx replied:
I think it's reasonable. I will probably move these into a separate folder called adasum under ops then.

olsaarik replied:
Yeah, that sounds good to me.

olsaarik committed:
Add some comments, more to be done

olsaarik commented on horovod/common/ops/adasum_operations.h:
```diff
@@ -0,0 +1,479 @@
+  // tag: a value used as the message tag for each send/recv in this algorithm. This is
+  //      useful for multithreaded scenarios. Remember to also create separate
+  //      reduction_comms instances when running with multiple threads.
+  // reduction_comms: TODO: Saeed please explain these.
```
@saeedmaleki could you add an explanation of what reduction_comms should be? (I don't understand VHDD well enough to explain them)

olsaarik commented on horovod/common/ops/mpi_p2p_operations.cc:
```diff
@@ -0,0 +1,147 @@
+  if (status != MPI_SUCCESS) {
+    throw std::logic_error("MPI_Allreduce failed, see MPI output for details.");
+  }
+}
```
Timeline events around this (and maybe the other MPI calls in this same file) would help with performance debugging of Adasum.

Commit:
changed license
addressed pr comments

Commit:
added timeline entries

Commit:
Merge branch 'alpha_light' of https://github.com/Tixxx/horovod into alpha_light

Commit:
fixed a bug where it return wrong number of gradients
removed redundant comments

Commit:
reduce number of allocations

Commit:
changed allocation order

olsaarik committed:
Add comments for Adasum VHDD reduction_comms

Commit:
decreased chunking size

olsaarik committed:
Add chunk size env var for adasum MPI

olsaarik committed:
Set default Adasum MPI chunk size to best value

For Intel MPI that is

Commit:
fixed test to use torch.cuda to set device
removed env variable
fixed the bug where stream allocation was not done properly

Commit:
Merge branch 'alpha_light' of https://github.com/Tixxx/horovod into alpha_light

Commit:
Merge branch 'alpha_light' of https://github.com/Tixxx/horovod into alpha_light_perf

Commit:
changed chunking to 1<<17

olsaarik committed:
Change default chunk size to be good for OpenMPI

olsaarik committed:
Merge pull request #24 from olsaarik/olsaarik/chunksizevar

Add chunk size environment variable

Commit:
Merge remote-tracking branch 'origin/alpha_light' into alpha_light_perf
preallocate host buffer for adasum

Commit:
removed chunk size variable in fusedallreduce

Commit:
Merge pull request #25 from Tixxx/alpha_light_perf

Alpha light perf

Commit:
Merge branch 'master' of https://github.com/Tixxx/horovod into alpha_light

Commit:
fixed integer overflow

Commit:
stabalizing

Commit:
fixed the stability issue

olsaarik committed:
Reorder functions in Adasum and make private what can be

olsaarik committed:
Undo stray change from another change set

Commit:
remove unnecessary CUDA_VISIBLE_DEVICES

Commit:
changed the cuda set device

saeedmaleki committed:
Merge pull request #26 from Tixxx/saemal/stability

Saemal/stability

olsaarik committed:
Merge branch 'alpha_light' into olsaarik/adasumreorder

olsaarik committed:
Remove duplicate function declaration

olsaarik committed:
Merge pull request #27 from olsaarik/olsaarik/adasumreorder

Reorder functions in adasum.h

olsaarik committed:
Delta approach support for TF DistributedOptimizer

Commit:
added delta optimizer for pytorch

Commit:
fixed a bug with precision

Commit:
removed wrapper type and made delta default for adasum

olsaarik committed:
Make delta approach default for adasum

Commit:
fixed bug where handles werent cleared

Commit:
changed backpass step logic and improved mnist example

Commit:
change logic of back pass count

Commit:
pr comments

olsaarik committed:
Merge branch 'olsaarik/tfdeltaoptimizer' into tix/torch_delta_optim

olsaarik committed:
Code cleanup and comments

olsaarik committed:
Fix TF delta allreduce call

olsaarik committed:
Fix for missing parameter

Commit:
added names for cache optimization

Commit:
Merge pull request #29 from Tixxx/tix/torch_delta_optim

Delta optimizers for both torch and tensorflow

Commit:
Merge branch 'master' into alpha_light

Commit:
improved resnet50 benchmark to use adasum

Commit:
improved delta optimizer to do asycn operations

Commit:
removed some un-used code

Commit:
Merge branch 'master' into alpha_light

Commit:
Merge branch 'alpha_light' of https://github.com/Tixxx/horovod into tix/delta_perf_hook

Commit:
removed apex support

Commit:
Merge pull request #30 from Tixxx/tix/delta_perf_hook

Tix/delta perf hook
